src/utilities/extendSchema.js

```diff
@@ -258,10 +264,103 @@ export function extendSchema(
     if (isUnionType(type)) {
       return extendUnionType(type);
     }
+    if (isEnumType(type)) {
+      return extendEnumType(type);
+    }
+    if (isInputObjectType(type)) {
+      return extendInputObjectType(type);
+    }
     // This type is not yet extendable.
     return type;
   }
 
+  function extendInputObjectType(
+    type: GraphQLInputObjectType,
+  ): GraphQLInputObjectType {
+    return new GraphQLInputObjectType({
+      name: type.name,
+      description: type.description,
+      fields: () => extendInputFieldMap(type),
+      astNode: type.astNode,
+    });
+  }
+
+  function extendInputFieldMap(type: GraphQLInputObjectType) {
+    const newFieldMap = Object.create(null);
+    const oldFieldMap = type.getFields();
+    Object.keys(oldFieldMap).forEach(fieldName => {
+      const field = oldFieldMap[fieldName];
+      newFieldMap[fieldName] = {
+        description: field.description,
+        type: extendFieldType(field.type),
+        defaultValue: field.defaultValue,
+        astNode: field.astNode,
+      };
+    });
+
+    // If there are any extensions to the fields, apply those here.
+    const extensions = typeExtensionsMap[type.name];
+    if (extensions) {
+      extensions.forEach(extension => {
+        extension.fields.forEach(field => {
+          const fieldName = field.name.value;
+          if (oldFieldMap[fieldName]) {
+            throw new GraphQLError(
+              `Field "${type.name}.${fieldName}" already exists in the ` +
+                'schema. It cannot also be defined in this type extension.',
+              [field],
+            );
+          }
+          newFieldMap[fieldName] = astBuilder.buildField(field);
+        });
+      });
+    }
+
+    return newFieldMap;
+  }
+
+  function extendEnumType(type: GraphQLEnumType): GraphQLEnumType {
+    return new GraphQLEnumType({
+      name: type.name,
+      values: extendValueMap(type),
+      astNode: type.astNode,
+    });
+  }
+
+  function extendValueMap(type: GraphQLEnumType) {
+    const oldValueMap = Object.create(null);
+    const newValueMap = Object.create(null);
+    type.getValues().forEach(value => {
+      newValueMap[value.name] = oldValueMap[value.name] = {
+        name: value.name,
+        description: value.description,
+        value: value.value,
```

Comment:
You shouldn't do that, because of this check:
https://github.com/graphql/graphql-js/blob/3cb450059d58f45dff95867a0674d7588be64933/src/type/definition.js#L1164
In the current version, it will convert all enum values without explicit `value` to `undefined`.
This is the very important case so please add a separate test for it.